osu.Server.Spectator/Hubs/ServerMultiplayerRoom.cs

```diff
@@ -27,13 +29,22 @@ public MatchTypeImplementation MatchTypeImplementation
             }
         }
 
-        public ServerMultiplayerRoom(long roomId, IMultiplayerServerMatchCallbacks hubCallbacks)
+        public readonly MultiplayerQueue QueueImplementation;
+
+        public ServerMultiplayerRoom(long roomId, IDatabaseFactory dbFactory, IMultiplayerServerMatchCallbacks hubCallbacks)
             : base(roomId)
         {
             this.hubCallbacks = hubCallbacks;
 
             // just to ensure non-null.
             matchTypeImplementation = createTypeImplementation(MatchType.HeadToHead);
+            QueueImplementation = new MultiplayerQueue(this, dbFactory, hubCallbacks);
```

Comment:
Hmm, don't really like the fact that `this` is passed down here. The `Initialise` method flow feels a bit sloppy the whole way down. Wondering if `Settings` could also be propagated down in a `Settings` setter rather than this method.

It would mean using a `new` or adding `virtual` to the base `MultiplayerRoom`, but interested in hearing thoughts.

Reply:
I guess the `MultiplayerRoom` is used in about 50 places so it's required to be passed. My main concern is how much cross-class operations which are going on here (the queue is doing writes to the room in many cases). Feels a bit confusing to follow.

In addition, the `ChangeMode` call takes in a new mode but is that even required if the settings are already available via the room reference? If it's to stay as-is, I'd probably want to see an assertion that the new queue mode matches the one in the room's settings.

Reply:
> Hmm, don't really like the fact that this is passed down here.

I'm following existing implementation (match types) which do the same thing.

> adding `virtual` to the base `MultiplayerRoom`

Completely against this. It's a model, not a business class. We've gone over this in the lead up to `ServerMultiplayerRoom`.

> It would mean using a `new`

A `new` what? `new MultiplayerSettings Settings { get; set; }`? It doesn't work for my purposes of `Initialise`, because I need async.

> the queue is doing writes to the room in many cases

In the same way as the match type implementations are doing writes to the room in many cases. I don't necessarily disagree because the match type implementations blew my mind the first time coming back to this project in... 4 months, but I disagree that this is something to be addressed here.

> the `ChangeMode` call takes in a new mode but is that even required if the settings are already available via the room reference

Again, also applicable to the match type implementation, I'm not doing anything new here. I will add an assert.

Reply:
Check whether the newest changes to `ChangeMode()` look better: https://github.com/ppy/osu-server-spectator/pull/84/commits/fed23800c923298f0e1990ba05b43c78efe83726

Reply:
Marginally I guess. Not a blocker either way, just reads a bit weird.